[online_image] Add JPEG support to online_image

guillempages opened the pull request:
# What does this implement/fix?

Add JPEG support to online_image, in addition to the already existing PNG support

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#4617

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
http_request:
  verify_ssl: false

online_image:
  - url: https://www.w3.org/MarkUp/Test/xhtml-print/20050519/tests/jpeg444.jpg
    id: jpeg_image
    format: JPEG
    type: RGB565

display:
  - platform: xyz
    # <...>
    lambda: |-
      it.image(0, 0, id(jpeg_image));
  
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

guillempages committed:
[online JPEG] Initial implementation

Add JPEG support on pyhton level, and empty implementation classes

Not working yet, but at least it compiles and runs

guillempages committed:
First working version

Small JPEGs can be downloaded and shown.

guillempages committed:
Add buffer resize possibility

guillempages committed:
Support big images

Add watchdog petting and chunk downloading, to avoid crashing
due to watchdog timeouts in case of huge images (>1MB)

guillempages committed:
Increase default buffer size

HeMan committed:
Update mdns for ESP-IDF (#8145)

clydebarrow commented on esphome/components/online_image/__init__.py:
```diff
@@ -69,11 +78,11 @@ def actions(self):
+        JPEGFormat(),
         PNGFormat(),
     )
 }
```
Allowing `format: jpg` as well will avoid some swearing.

```suggestion
}
IMAGE_FORMATS.update({"JPG": IMAGE_FORMATS["JPEG"]})
```

guillempages replied:
Done. I've also added a test for that.

clydebarrow commented on esphome/components/online_image/online_image.h:
```diff
@@ -119,6 +126,12 @@ class OnlineImage : public PollingComponent,
   DownloadBuffer download_buffer_;
+  /**
+   * This is the *initial* size of the download buffer, not the current size.
+   * The download buffer can be resized at runtime; the download_buffer_size_ will
```
```suggestion
   * The download buffer can be resized at runtime; the download_buffer_initial_size_ will
```

Maybe the constructor parameter should also be renamed?

guillempages replied:
Thanks for finding this one.

W.r.t. renaming the constructor parameter; I'd rather not. The constructor parameter is actually used to set the size of the download buffer (as the name says); just in case of Jpeg decoder, the image needs to be fully downloaded before decoding (because of the way the decoder works), so the initial size value is still needed later on.

bdraco committed:
Include Bluetooth connection slot allocations in connections free message (#8148)

Rapsssito committed:
[esp32_ble_server] Create custom services, characteristics and descriptors (#7009)

Co-authored-by: Clyde Stubbs <2366188+clydebarrow@users.noreply.github.com>

guillempages committed:
Review comments

 - support "JPG"
 - update variable documentation

  added an additional test for JPG as well

guillempages committed:
Merge branch 'dev' into online_image_jpeg

guillempages committed:
Add additional checks for unsupported formats

Progressive JPEGs are not being decoded right; avoid decoding them
at all and show an error.

guillempages committed:
Reorder progressive check
